src/govuk/components/character-count/template.njk

```diff
@@ -1,10 +1,18 @@
+{% from "../../macros/i18n.njk" import govukPluralisedI18nAttributes %}
+
 {% from "../textarea/macro.njk" import govukTextarea %}
 {% from "../hint/macro.njk" import govukHint %}
 
 <div class="govuk-character-count" data-module="govuk-character-count"
 {%- if params.maxlength %} data-maxlength="{{ params.maxlength }}"{% endif %}
 {%- if params.threshold %} data-threshold="{{ params.threshold }}"{% endif %}
-{%- if params.maxwords %} data-maxwords="{{ params.maxwords }}"{% endif %}>
+{%- if params.maxwords %} data-maxwords="{{ params.maxwords }}"{% endif %}
+{%- if params.charactersUnderLimitHtml %}{{govukPluralisedI18nAttributes('characters-under-limit', params.charactersUnderLimitHtml)}}{% endif %}
```

Comment:
Super quick not-even-tried-it example, but how about a function that does all i18n attributes?

```js
{% set i18n = params.i18n %}
```

```js
{%- if i18n %}{{ govukI18nAttributes([
  ['characters-under-limit', i18n.charactersUnderLimitHtml],
  ['characters-at-limit', i18n.charactersAtLimitHtml],
  ['characters-over-limit', i18n.charactersOverLimitHtml],
  ['words-under-limit', i18n.wordsUnderLimitHtml],
  ['words-at-limit', i18n.wordsAtLimitHtml],
  ['words-over-limit', i18n.wordsOverLimitHtml]
])}}{% endif %}
```

Reply:
The idea is appealing, and we don't really need to introduce a separate `i18n` namespace for them, we can just pick the values from `params` directly as we do have to read them manually (unless someone knows of some black magic to go kebab-case to cameCase or the other way around with Nunjucks without adding a filter).

There's a little trickiness to it as translations can be strings or object, depending of if they're pluralised or not. Using the current helper, the template makes it clear which attribute expects which, so we'd need to devise an API that lets us do the same, maybe:

```html
{{ govukI18nAttributes({
  'characters-under-limit': {plurals: {params.charactersUnderLimitHtml}},
  'characters-over-limit': {plurals: {params.charactersOverLimitHtml}},
  'characters-at-limit': {string: params.charactersAtLimitHtml},
  'words-under-limit': {plurals: {params.wordsUnderLimitHtml}},
  'words-over-limit': {plurals: {params.wordsOverLimitHtml}},
  'words-at-limit': {string: params.wordsAtLimitHtml}
 })}}
```

That said, I think there's ground for a codebase wise look at some abstraction around data attributes, like a `govukDataAttributes` helper on a similar model, as setting them is always the same thing:
- check if the value is truthy (or not undefined in a few cases)
- set `data-{{name}}={{value}}` (sometimes escaped)

Reply:
Added an issue to track: https://github.com/alphagov/govuk-frontend/issues/2917 😄

Reply:
@romaricpascal Yeah, look less at the "how" and more of a nice neat way to generate attributes

We already have this syntax elsewhere for `.attributes`:

```js
{%- for attribute, value in message.attributes %} {{attribute}}="{{value}}"{% endfor %}
```

Then other "ad-hoc" data attributes in places:

```js
{%- if params.disableAutoFocus !== undefined %} data-disable-auto-focus="{{ params.disableAutoFocus }}"{% endif %}
```

If each template had a map (or object) of all its available attributes, defaults, parameter overrides etc, you could write both regular and data attributes into the template in a nice unit-testable way

This certainly isn't a blocker, just an opportunity to makes the template look neater